xir/program.py

```diff
@@ -137,27 +137,55 @@ def use_floats(self) -> bool:
         return self._use_floats
 
 
+class ObservableFactor:
+    """Observable factor to be used in observable definitions.
+
+    Args:
+        name (str): the name of the factor
+        params ([Any]): the parameters of the factor
+        wires ([int]): the wires this factor acts on
+    """
+
+    def __init__(self, name: str, params: Params, wires: Sequence[Wire]) -> None:
+        self.name = name
+        self.params = list(params or [])
+        self.wires = tuple(wires or [])
+
+    def __str__(self) -> str:
+        wires = ", ".join([str(v) for v in self.wires])
+        if len(self.params) == 0:
+            return f"{self.name}[{wires}]"
+        else:
+            params = ", ".join([str(v) for v in self.params])
+            return f"{self.name}({params})[{wires}]"
```

Comment:
The logic here is fine although I have a few code style suggestions:
1. A clean usage of `map` should be preferred to list comprehensions. For example, compare
    ```python
    wires = ", ".join([str(v) for v in self.wires])
    ```
    to
    ```python
    wires = ", ".join(map(str, self.wires))
    ```
2. It is more Pythonic to check if a sequence is empty using
    ```python
    if self.params:
    ```
3. There is no need to use `else` when the final line of an `if` block is `return`.

Putting it all together,

```suggestion
        wires = ", ".join(map(str, self.wires))
        if self.params:
            return f"{self.name}[{wires}]"

        params = ", ".join(map(str, self.params))
        return f"{self.name}({params})[{wires}]"
```